File: leetcode/snakes-and-ladders.py
from typing import List


class Solution:
    def snakesAndLadders(self, board: List[List[int]]) -> int:
        queue = []
        visited = {}
        n = len(board)
        queue.append((1, 0))
        while len(queue) != 0:
            current, current_depth = queue.pop(0)  # TODO Dequeue로 바꾸면 조금 나음
            if current in visited:
                continue
            else:
                visited[current] = True
            if current == n * n:
                return current_depth
            next = self.getNext(current, n)
            for i in range(next[0], next[1]+1):
                co = self.getCoordinate(i, n)
                if board[co[0]][co[1]] != -1:
                    next_co = board[co[0]][co[1]]
                    if not next_co in visited:
                        queue.append((next_co, current_depth+1))
                else:
                    if not i in visited:
                        queue.append((i, current_depth+1))
        return -1

    def snakesAndLaddersWrongButConcept(self, board: List[List[int]]) -> int:
        queue = []
        n = len(board)
        queue.append(1)
        cnt = -1
        while len(queue) != 0:
            cnt += 1
            current = queue.pop(0)  # TODO Dequeue로 바꾸면 조금 나음
            print('current: ', current)
            if current == n * n:
                return cnt
            next = self.getNext(current, n)
            flag = False
            # check final
            if next[1] == n*n:
                queue.append(n*n)
                continue
            for i in range(next[0], next[1]+1):
                co = self.getCoordinate(i, n)
                print('compare: ', co, board[co[0]][co[1]])
                if board[co[0]][co[1]] != -1:
                    next_co = board[co[0]][co[1]]
                    print('current:', self.getCoordinate(
                        current, n), ' to next:', co, '. so that can use ladder or snake.', self.getCoordinate(next_co, n))
                    queue.append(next_co)
                    flag = True
                    break
            if not flag:
                i = self.getNext(current, n)[1]
                print('current:', self.getCoordinate(
                    current, n), ' to next:', self.getCoordinate(i, n))
                queue.append(i)
        return cnt

    def getNext(self, curr: int, n: int) -> [int, int]:
        return [curr + 1, min(curr + 6, n*n)]

    def getCoordinate(self, cur: int, n: int) -> (int, int):
        row = (n*n - cur) // n
        col = ((cur - 1) % n)
        if (n-row) % 2 == 0:
            col = n - 1 - col
        return (row, col)


tcs = [
    [[-1, -1, -1, -1, -1, -1], [-1, -1, -1, -1, -1, -1], [-1, -1, -1, -1, -1, -1],
     [-1, 35, -1, -1, 13, -1], [-1, -1, -1, -1, -1, -1], [-1, 15, -1, -1, -1, -1]],
    [[-1, -1, -1], [-1, 9, 8], [-1, 8, 9]],
    [[-1, -1],
     [-1, 3]],
    [[1, 1, -1], [1, 1, 1], [-1, 1, 1]]
]


# sol = Solution().getCoordinate(9, 3)
for tc in tcs:
    sol = Solution().snakesAndLadders(tc)
    print(sol)

File: leetcode/snakesAndLadders.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Set;
import java.util.TreeSet;

public class snakesAndLadders {
    public static void main(String[] args) {
        System.out.println("Hello World");
//        int[][] given = {
//                {1, 1, -1},
//                {1, 1, 1,},
//                {-1, 1, 1}
//        };
        int[][] given = {
                {-1, -1, 19, 10, -1},
                {2, -1, -1, 6, -1},
                {-1, 17, -1, 19, -1},
                {25, -1, 20, -1, -1},
                {-1, -1, -1, -1, 15}
        };

//        int[][] given = {
//                {-1, -1},
//                {-1, 3},
//        };
//        int[][] given = {
//                {-1, -1, -1},
//                {-1, 9, 8},
//                {-1, 8, 9}
//        };
//        int[][] given = {
//                {-1, -1, -1, -1, -1, -1},
//                {-1, -1, -1, -1, -1, -1},
//                {-1, -1, -1, -1, -1, -1},
//                {-1, 35, -1, -1, 13, -1},
//                {-1, -1, -1, -1, -1, -1},
//                {-1, 15, -1, -1, -1, -1}
//        };
        MySnakeAndLadders solution = new MySnakeAndLadders();
        System.out.println(solution.solve(given));
    }


}

class MySnakeAndLadders {
    class Pair {
        int Pos;
        int Depth;

        public Pair(int pos, int depth) {
            Pos = pos;
            Depth = depth;
        }
    }

    public int solve(int[][] board) {
        int n = board.length;
        // 1부터 시작
        Queue<Pair> queue = new LinkedList<>();
        Set<Integer> visited = new TreeSet<>();
        queue.add(new Pair(1, 0));
        // queue
        while (!queue.isEmpty()) {
            Pair current = queue.remove();
            if (visited.contains(current.Pos)) {
                continue;
            }
            visited.add(current.Pos);
            System.out.println("Current: " + current.Pos + ", " + current.Depth);
            if (current.Pos == n * n) {
                return current.Depth;
            }
            int[] next = getNext(current.Pos, n);
            if (next[1] == n * n) {
                queue.add(new Pair(next[1], current.Depth + 1));
                continue;
            }
            for (int i = next[0]; i <= next[1]; i++) {
                int[] coord = currentToCoordinate(i, n);
                if (board[coord[0]][coord[1]] != -1) {
                    queue.add(new Pair(board[coord[0]][coord[1]], current.Depth + 1));
                } else {
                    queue.add(new Pair(i, current.Depth + 1));

                }
            }
        }
        return -1;
    }

    public int[] currentToCoordinate(int curr, int n) {
        int row = (n - 1) - (curr - 1) / n;
        int col = (curr - 1) % n;
        if (((curr - 1) / n) % 2 != 0) {
            col = n - 1 - col;
        }
        return new int[]{row, col};
    }

    public int[] getNext(int curr, int n) {
        int start = curr + 1;
        int end = curr + 6 < n * n ? curr + 6 : n * n;
        return new int[]{start, end};
    }

}
